### ISSUE TITLE
Fresnel MFT

### ISSUE THREAD

**Issue by Jordan-Dennis** [LouisDesdoigts/dLux#86]
Hi all,
Thought I would create a separate issue for this. Essentially, we want to use the `_mft` in the `GaussianPropagator` in the long term to avoid sampling difficulties and increase the speed. As a part of this we need to specify the size of the output region in units of $\lambda/D$ where $D$ is the width of the pupil. I may be confusing myself but I believe at any point in the optical propagation we can use, $2 \times \omega_{z} / (\lambda / D)$, where $\omega_{z}$ is the beam radius at $z$. 

Regards

Jordan

**Comment by benjaminpope** [comment]
The actual matrix FT doesn't know anything about angular or cartesian coordinates - it is just a mathematical operator. Its Fourier exponents are expressed in terms of λ/D because these are the convenient units, but it's really just about establishing where each Fourier component sits in relation to the Nyquist frequency.

In morphine (see [notebook](https://github.com/benjaminpope/morphine/blob/stable/notebooks/morphine_uv.ipynb)), I did a crude hack of the `poppy` matrix DFT to instead generate `u, v` plane (ie spatial frequency) Fourier transforms of the image: after some head scratching I arrived at the function 

```python
def calc_uv(wavefront,opsys,npix=128):
    nlamd = arcsec2rad*wavefront.pixelscale * wavefront.shape[0] / WAVELENGTH*2*opsys.planes[0].pupil_diam
    pupil = wavefront.intensity
    return minimal_dft(pupil, nlamd, npix)
```

to 'fake it' for the DFT expressed in units of λ/D so that it would instead give correctly normalized spatial frequencies. This trick, though not this numerical coefficient, should apply to angular and cartesian FTs we need to do, if we can find the right expression for the output frequencies with a bit of pen and paper / inspecting `np.fft.fftfreq` / experimental mathematics.

There is nothing to stop us from simply re-expressing the FFT in Fresnel in terms of an appropriately hacked matrix FT, and then we can think about doing things like the 2-sided MFT (issue https://github.com/LouisDesdoigts/dLux/issues/82) to accelerate it, or rescale the windowing, et cetera, in a way that FFTs can't do without a lot of padding.

**Comment by benjaminpope** [comment]
Fresnel still having problems: minimal working example.

Usual imports:

```python
import numpy as onp
import matplotlib.pyplot as plt
import matplotlib as mpl

import jax
from jax.config import config
config.update("jax_enable_x64", True)
import jax.numpy as np

import equinox as eqx
import optax

import astropy.units as u 

import functools as ft # used for filtering
from tqdm import tqdm

import poppy 

plt.rcParams['figure.dpi'] = 120
plt.rcParams['image.cmap'] = 'hot'
plt.rcParams["text.usetex"] = 'true'

from dLux.base import *
from dLux.layers import *
from dLux.propagators import *

import dLux as dl```

Generate an HST model:
```python
wl = 1.5e-6 # m
aperture = 2.4 # m
wf_npix = 512       

diam = 2.4 * u.m
fl_pri = 5.52085 * u.m
```

Now we set up an HST model from poppy
```python
hstf = poppy.FresnelOpticalSystem(pupil_diameter=2.4*u.m, npix=wf_npix, beam_ratio=0.25)
m1 = poppy.QuadraticLens(fl_pri, name='Primary')

hst_aper = poppy.CompoundAnalyticOptic(opticslist=[
    poppy.SecondaryObscuration(secondary_radius=0.3, n_supports=4, support_width=0.1),
    poppy.CircularAperture(radius=aperture/2)])

hstf.add_optic(hst_aper)
hstf.add_optic(m1)

hstf.add_optic(poppy.ScalarTransmission(planetype=poppy.fresnel.PlaneType.image, name='focus'), 
               distance=fl_pri-10*u.um);
```

Generate a poppy PSF with fresnel
```python
# Generate poppy PSF and get the aperture array to apply to our mode
hst_fits, intermediates = hstf.calc_psf(wavelength=wl, display_intermediates=True, return_intermediates=True)
hst_psf_p = np.flipud(np.flipud(hst_fits[0].data).T) # Flip to match ∂Lux
hst_ap = np.array(intermediates[0].amplitude) # Take aperture for ∂Lux

osamp = intermediates[-1].oversample
det_npix = hst_psf_p.shape[0]
```

now try ∂Lux:
```python
# Define approximate fl and pixelsize, we will optimise the fl to match poppy

fl = 5.52085  # approximate fl, to be inferred
det_pixsize = hst_fits[0].header['PIXELSCL']  # fixed: 0.03222887597600081 arcsec/pix

hst_layers = [
    CreateWavefront(wf_npix*4, wf_npix*4),
    ApplyAperture(hst_ap),
    NormaliseWavefront(),
    PhysicalFresnel(pixels_out=det_npix, focal_length=fl, focal_shift=-1e-6,
        pixel_scale_out=det_pixsize, inverse=False)]

wavels = np.array([wl])
hst = OpticalSystem(hst_layers, wavels=wavels)
psf = hst()
```

The PhysicalFresnel prop also prints a lot of metadata, which I assume was for debugging, but which will be bad in an optimization loop... 

```
initial:  1.0000000000000007
thin lens:  Traced<ShapedArray(float64[])>with<BatchTrace(level=2/0)> with
  val = DeviceArray([1.], dtype=float64)
  batch_dim = 0
rho_1:  Traced<ShapedArray(float64[])>with<BatchTrace(level=2/0)> with
  val = DeviceArray([1.], dtype=float64)
  batch_dim = 0
_propagate: Traced<ShapedArray(float64[])>with<BatchTrace(level=2/0)> with
  val = Traced<ShapedArray(float64[1])>with<BatchTrace(level=1/0)> with
    val = DeviceArray([[6.49630424e+13]], dtype=float64)
    batch_dim = 0
  batch_dim = 0
pixel_scale: Traced<ShapedArray(float64[])>with<BatchTrace(level=2/0)> with
  val = Traced<ShapedArray(float64[1])>with<BatchTrace(level=1/0)> with
    val = DeviceArray([[6.49630424e+13]], dtype=float64)
    batch_dim = 0
  batch_dim = 0
transfer_function:  Traced<ShapedArray(float64[])>with<BatchTrace(level=2/0)> with
  val = Traced<ShapedArray(float64[1])>with<BatchTrace(level=1/0)> with
    val = DeviceArray([[9.47266328e+23]], dtype=float64)
    batch_dim = 0
  batch_dim = 0
rho_2:  Traced<ShapedArray(float64[])>with<BatchTrace(level=2/0)> with
  val = Traced<ShapedArray(float64[1])>with<BatchTrace(level=1/0)> with
    val = DeviceArray([[9.47266328e+23]], dtype=float64)
    batch_dim = 0
  batch_dim = 0
```

Then we get a chequerboarded result: I think this is the same bug we fixed the other day about fftshift, but perhaps not committed?

![image](https://user-images.githubusercontent.com/6024213/180784557-8206665d-7558-4412-8010-9b612bd2eeb6.png)

**Comment by Jordan-Dennis** [comment]
Yes sorry,
The working code has not been pushed to `GitHub` yet. That is something that I want to do today. This issue has been fixed.

**Comment by LouisDesdoigts** [comment]
Yeah you're mixing a few different things there that shouldn't be mixed. Physical propagators need Physical wavefronts, which define all units in m/pix and therefore require focal lengths, but you've popped in an angular pixel size. 

I've got a working version of AngularWavefront on my JWST_dev branch which I might try and merge to main tonight. Using angular wavefront has made life easier for matching poppy because it by default uses angular units.

Also have you done #60 @Jordan-Dennis? I assume so if youre using  GaussianWavefront...

**Comment by Jordan-Dennis** [comment]
Actually no I have been working with the exposed internals for this. I will try and find time on the weekend to experiment with it.

**Comment by LouisDesdoigts** [comment]
Ah yes, I actually though I had solved this on my branch, but I'm just using a PhysicalWavefront with and AngularMFT which works fine because its just a different output unit. 

Yeah because we want to merge the Fresnel stuff we should probably raise this on the priority list. As a work-around we could just pass wavelength and offset in the params_dict and create the wavefront entirely inside CreateWavefront. I might do this as a work around until we get proper type promotion. Will re-post this in #60.

**Comment by LouisDesdoigts** [comment]
So just reviving this a bit, I think I know a not-too-hard workable solution, assuming that we know all the physical pixel scales along the propagation path (which I believe to be the case). 

FFT's naturally sample at 1 pixel per fringe (PPF), but MFT's have a variable PPF output. Exactly replacing FFTs with MFTs is trivial, as number of fringes is npix/2 (since its measured from the center). Given that we know the exact pixel scale at each point in the propagation with the FFT algorithms, and we want to use MFTs to specify our output pixel scale in physical (not fringe) units, we only need to know the ratio of the native output pixel scale to the desired output pixel scale. Given that, number of fringes is just that ratio multiplied by the native npix/2 number of fringes output. 

mathematically this would look like this:

native sampling = $\alpha$ = $\text{npix}/2$ (fringes)
native pixel scale = $\delta_{n}(\lambda, D)$ (m/pix) = ... output from FFT algorithm, which contains our wavelength dependence
desired pixel scale = $\delta_{d}$ (m/pix)
fringe sampling out = $\alpha$ * $\delta_{n}(\lambda, D)$ / $\delta_{d}$ (fringes)

Then we should be able to feed this fringe sampling out value as the 'number of fringes' parameter to the MFT! (assuming I've got the fraction the right way around) Let me know if you think this would work @Jordan-Dennis. Its obviously not an optical derivation of the desired quantity, but if it works 🤷‍♂️. Maybe in the long term we can reverse engineer the derivation, for now I think we just want it working.

**Comment by Jordan-Dennis** [comment]
Hey @LouisDesdoigts, could you please update us on how this went. I understand that you got it working specifically for the plane to plane mode of the propagation.

**Comment by LouisDesdoigts** [comment]
A new version of this functionality is being re-build elsewhere, and most of these issues and discussion is no longer relevant, so will close this for now.
